File: app/src/test/java/openfoodfacts/github/scrachx/openfood/utils/UtilsTest.java
package openfoodfacts.github.scrachx.openfood.utils;

import android.content.Context;
import android.content.pm.PackageManager;

import org.junit.Test;

import java.util.Locale;

import openfoodfacts.github.scrachx.openfood.R;
import openfoodfacts.github.scrachx.openfood.models.Product;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class UtilsTest {
    @Test
    public void getImageGrade() {
        final Product mockProduct = mock(Product.class);

        when(mockProduct.getNutritionGradeFr()).thenReturn("a");
        assertThat(Utils.getImageGrade(mockProduct)).isEqualTo(R.drawable.ic_nutriscore_a);

        when(mockProduct.getNutritionGradeFr()).thenReturn("b");
        assertThat(Utils.getImageGrade(mockProduct)).isEqualTo(R.drawable.ic_nutriscore_b);

        when(mockProduct.getNutritionGradeFr()).thenReturn("c");
        assertThat(Utils.getImageGrade(mockProduct)).isEqualTo(R.drawable.ic_nutriscore_c);

        when(mockProduct.getNutritionGradeFr()).thenReturn("d");
        assertThat(Utils.getImageGrade(mockProduct)).isEqualTo(R.drawable.ic_nutriscore_d);

        when(mockProduct.getNutritionGradeFr()).thenReturn("e");
        assertThat(Utils.getImageGrade(mockProduct)).isEqualTo(R.drawable.ic_nutriscore_e);

        when(mockProduct.getNutritionGradeFr()).thenReturn("");
        assertThat(Utils.getImageGrade(mockProduct)).isEqualTo(Utils.NO_DRAWABLE_RESOURCE);

        when(mockProduct.getNutritionGradeFr()).thenReturn(null);
        assertThat(Utils.getImageGrade(mockProduct)).isEqualTo(Utils.NO_DRAWABLE_RESOURCE);
    }

    @Test
    public void getRoundNumber() {
        /* TODO: Fix method before testing
        assertThat(Utils.getRoundNumber("")).isEqualTo("?");
        assertThat(Utils.getRoundNumber(null)).isEqualTo("?");
        // TODO: Something for string
        // assertThat(Utils.getRoundNumber("test")).isEqualTo("?");
        assertThat(Utils.getRoundNumber("0")).isEqualTo("0");
        assertThat(Utils.getRoundNumber(0)).isEqualTo("0");
        assertThat(Utils.getRoundNumber(1)).isEqualTo("1");
        assertThat(Utils.getRoundNumber(1.7f)).isEqualTo("1.7");
        assertThat(Utils.getRoundNumber(1.75f)).isEqualTo("1.75");
        assertThat(Utils.getRoundNumber(1.754f)).isEqualTo("1.75");
        assertThat(Utils.getRoundNumber(1.756f)).isEqualTo("1.76");
        assertThat(Utils.getRoundNumber(1.756f)).isEqualTo("1.76");

         */
    }

    @Test
    public void isHardwareCameraInstalled() {
        final Context mockCtx = mock(Context.class);
        final PackageManager mockPM = mock(PackageManager.class);

        when(mockCtx.getPackageManager()).thenReturn(mockPM);

        // Test for best condition
        when(mockPM.hasSystemFeature(PackageManager.FEATURE_CAMERA))
            .thenReturn(true);
        assertThat(Utils.isHardwareCameraInstalled(mockCtx)).isTrue();

        // False condition
        when(mockPM.hasSystemFeature(PackageManager.FEATURE_CAMERA))
            .thenReturn(false);

        assertThat(Utils.isHardwareCameraInstalled(mockCtx)).isFalse();
    }

    @Test
    public void getServingInOz_from_ml() {

        assertThat(UnitUtils.getServingInOz("100 ml"))
            .isEqualTo(String.format(Locale.getDefault(), "%.2f", 3.38) + " oz");
    }

    @Test
    public void getServingInOz_from_cl() {
        assertThat(UnitUtils.getServingInOz("250 cl"))
            .isEqualTo(String.format(Locale.getDefault(), "%.2f", 84.53) + " oz");
    }

    @Test
    public void getServingInOz_from_l() {

        assertThat(UnitUtils.getServingInOz("3 l"))
            .isEqualTo(String.format(Locale.getDefault(), "%.2f", 101.44) + " oz");
    }
}

File: app/src/test/java/openfoodfacts/github/scrachx/openfood/utils/UtilsTest.kt
package openfoodfacts.github.scrachx.openfood.utils

import android.content.Context
import android.content.pm.PackageManager
import com.google.common.truth.Truth.assertThat
import openfoodfacts.github.scrachx.openfood.R
import openfoodfacts.github.scrachx.openfood.models.Product
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import java.util.*

class UtilsTest {
    @Test
    fun testGetImageGrade() {
        val mockProduct = mock(Product::class.java)

        whenever(mockProduct.nutritionGradeFr) doReturn "a"
        assertThat(mockProduct.getNutriScoreResource()).isEqualTo(R.drawable.ic_nutriscore_a)

        whenever(mockProduct.nutritionGradeFr) doReturn "b"
        assertThat(mockProduct.getNutriScoreResource()).isEqualTo(R.drawable.ic_nutriscore_b)

        whenever(mockProduct.nutritionGradeFr) doReturn "c"
        assertThat(mockProduct.getNutriScoreResource()).isEqualTo(R.drawable.ic_nutriscore_c)

        whenever(mockProduct.nutritionGradeFr) doReturn "d"
        assertThat(mockProduct.getNutriScoreResource()).isEqualTo(R.drawable.ic_nutriscore_d)

        whenever(mockProduct.nutritionGradeFr) doReturn "e"
        assertThat(mockProduct.getNutriScoreResource()).isEqualTo(R.drawable.ic_nutriscore_e)

        whenever(mockProduct.nutritionGradeFr) doReturn ""
        assertThat(mockProduct.getNutriScoreResource()).isEqualTo(R.drawable.ic_nutriscore_unknown)

        whenever(mockProduct.nutritionGradeFr) doReturn null
        assertThat(mockProduct.getNutriScoreResource()).isEqualTo(R.drawable.ic_nutriscore_unknown)
    }

    @Test
    fun getRoundNumber() {
        assertThat("".getRoundNumber()).isEqualTo("?")
        assertThat("test".getRoundNumber()).isEqualTo("?")
        assertThat("0".getRoundNumber()).isEqualTo("0")
        assertThat(0.0f.getRoundNumber()).isEqualTo("0")
        assertThat(1.00f.getRoundNumber()).isEqualTo("1")
        assertThat(1.70f.getRoundNumber(Locale.ENGLISH)).isEqualTo("1.7")
        assertThat(1.75f.getRoundNumber(Locale.ENGLISH)).isEqualTo("1.75")
        assertThat(1.754f.getRoundNumber(Locale.ENGLISH)).isEqualTo("1.75")
        assertThat(1.756f.getRoundNumber(Locale.ENGLISH)).isEqualTo("1.76")
        assertThat(1.756f.getRoundNumber(Locale.ENGLISH)).isEqualTo("1.76")
    }

    @Test
    fun isHardwareCameraInstalled() {
        val mockPM = mock<PackageManager> {}
        val mockCtx = mock<Context> {
            on { this.packageManager } doReturn mockPM
        }

        // Test for true/false
        whenever(mockPM.hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY)) doReturn true
        assertThat(isHardwareCameraInstalled(mockCtx)).isTrue()

        whenever(mockPM.hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY)) doReturn false
        assertThat(isHardwareCameraInstalled(mockCtx)).isFalse()
    }

}